### ISSUE TITLE
Provide a local, secure configuration for inclusion in Homebrew

### ISSUE THREAD

**Issue by log-to-code** [0xERR0R/blocky#839]

**Comment by ThinkChaos** [comment]
I agree that changing the default from all hosts to localhost only would be nice security wise.
The docker image should keep the all hosts by default though since it container runtimes have a firewall by default it's not a security issue and helps with usability.

@0xERR0R any thoughts on changing this?  
IMO it's a good moment to consider this since the development branch already has backwards incompatible config changes from #771.

**Comment by 0xERR0R** [comment]
I think the default behavior should be: blocky listens on all interfaces, since this is a common use case. We can provide some example configuration files for different usage scenarios, but it is not so easy:

- Which upstream servers should we use? "8.8.8.8" and "1.1.1.1" are fast but some people say, "oh no google is evil, we need more privacy". Some people say "we can't take plain DNS because it is insecure and/or my provider blocky traffic on port 53". We can take DoH servers from some free projects, but they are slower and the speed is depending on users location.
- Which blocklists/whitelists should we use? Without any blocklist - I think most uses want to block some ads. With some small ad-blocking list - ok, could be good enough, but what is about tracker? Some users use facebook, some users hate facebook. Facebook aps have problems if some facebook domains are blocked...

Also relevant Issue: https://github.com/0xERR0R/blocky/issues/380

**Comment by ThinkChaos** [comment]
Since you mention it, as part of #380, we'd need to change the default port to be >1024. Which is also an uncommon config, but a better default.  
I think of listening on localhost similarly: it's a less common config, but as a default it's more common (all distribs/package managers I know do this) and more secure. Secure by default is pretty important IMO and users should opt-in to opening the server to other machines by specifying IPs in the config.

I agree choosing defaults for upstreams and blocklists is not possible.  
What if we set the defaults be conservative (listen on `localhost:5353`, upstream=system resolver, no blocklists, etc.), and rely on the example config to provide something that covers the more common use cases?  
Most users will probably copy the whole example config to get started, so it wouldn't add any overhead for them. And the defaults would work out of the box, be secure and non opinionated.

**Comment by log-to-code** [comment]
> Which blocklists/whitelists should we use?

What about deferring to the list of DoH providers included in Firefox?

**Comment by log-to-code** [comment]
> Which blocklists/whitelists should we use?

These are tough decisions so I recommend keeping the scope of the Homebrew configuration to providing just enough config to get blocky running as a replacement DNS resolver on a MacOS device without exposing the service to the outside world. Then to help users make full use of the capabilities of blocky, link to the example config at docs/config.yml at the top of the config file as a comment and provide commented-out blocklist/allowlist sections with hints like "If you would like to set up adblocking, find adblocking lists on the internet and add them here."

**Comment by 0xERR0R** [comment]
> Since you mention it, as part of #380, we'd need to change the default port to be >1024. Which is also an uncommon config, but a better default. I think of listening on localhost similarly: it's a less common config, but as a default it's more common (all distribs/package managers I know do this) and more secure. Secure by default is pretty important IMO and users should opt-in to opening the server to other machines by specifying IPs in the config.
> 

Why do you want to change the default port to >1024? Docker image runs already with non-root user and port 53. For binary installation user can use setcap.
From security point of view, I agree, config should be more restrictive. But I think the default config should just run in the most cases. The common use case is to run blocky on a network device on local network and in this case it should listen for incoming connections. So IMHO, I would be better to leave the default config (listen to all interfaces) and user should restrict it for special use cases (for example using blocky on a local machine).

We can provide different config files and link them with a use case description on the documentation page, what do you think?

**Comment by log-to-code** [comment]
Anyone with a home network of special purpose devices will be expecting blocky to listen to all hosts and likely configuring their instance of blocky for adblocking, logging, metrics, etc. Same with cloud. However a regular MacBook user, maybe a beginning developer, installing blocky as a local adblocker shouldn't be left exposed to the world. A malicious actor making inappropriate or sensitive requests to your laptop isn't that much of a security issue but leaving an inexperienced user vulnerable to that doesn't seem right.

**Comment by ThinkChaos** [comment]
> We can provide different config files and link them with a use case description on the documentation page, what do you think?

Your call.  
But for the record, I think this will cause more work both here (maintain more configs) and downstream (override defaults), and see it as just a matter of time before there's a vulnerability in blocky (counting dependencies) that is exposed by default.

> For binary installation user can use setcap.

That's an extra step that requires root, so IMO is not valid for "blocky should run without config".

**Comment by log-to-code** [comment]
A new pull request has been opened to include blocky in homebrew-core because the last one was closed.

https://github.com/Homebrew/homebrew-core/pull/127041

**Comment by log-to-code** [comment]
blocky is now available in Homebrew. The example config was sufficient.

**Comment by log-to-code** [comment]
https://formulae.brew.sh/formula/blocky
